### ISSUE TITLE
[no-misused-signals] restrict incorrect usage of signals - proposal

### ISSUE THREAD

**Issue by MillerSvt** [angular-eslint/angular-eslint#1891]
**Description of the proposal**

This rule forbids providing Signals to logical locations such as if statements in places where the TypeScript compiler allows them but they are not handled properly. These situations can often arise due to a missing call or just a misunderstanding of the way signals are handled.

Reference: https://typescript-eslint.io/rules/no-misused-promises/

**Incorrect code**
```typescript
const signal = signal(0);

if (signal) {
  // Do something
}

const val = signal ? 123 : 456;

while (signal) {
  // Do something
}
```
```html
@if (signal) {
   something  
}

@switch (signal) {
   something
}
```

**Correct code**
```typescript
const signal = signal(0);

if (signal()) {
  // Do something
}

const val = signal() ? 123 : 456;

while (signal()) {
  // Do something
}
```
```html
@if (signal()) {
   something  
}

@switch (signal()) {
   something
}
```

**Comment by MillerSvt** [comment]
I can do this. Wait for approve.

**Comment by PowerKiKi** [comment]
This seems to be a duplicate of #1753 and should probably be merged into a single issue.

**Comment by marcorohn** [comment]
Since signals are functions, this is also valid typescript code, but will never run, as `Function.length` always returns `0` for signals:

```typescript
const mySignal = signal(['hello', 'world']);
if (mySignal.length > 0) {
  ...
}
```

Just happened to me recently when converting a component to use signals, and I'm recently sure that is not that uncommon

**Comment by JamesHenry** [comment]
Thanks folks, it makes sense to me and I'll happily accept a PR for this rule
